File: subgraphs/inventory/otel.js
const { NodeSDK } = require('@opentelemetry/sdk-node');
const { getNodeAutoInstrumentations } = require('@opentelemetry/auto-instrumentations-node');
const { OTLPTraceExporter } = require('@opentelemetry/exporter-trace-otlp-http');
const { OTLPMetricExporter } = require('@opentelemetry/exporter-metrics-otlp-http');
const { PeriodicExportingMetricReader } = require('@opentelemetry/sdk-metrics');
const { Resource } = require('@opentelemetry/resources');
const { ATTR_SERVICE_NAME, ATTR_SERVICE_VERSION } = require('@opentelemetry/semantic-conventions');
const { W3CTraceContextPropagator } = require('@opentelemetry/core');

/**
 * Initialize OpenTelemetry instrumentation for a subgraph service
 * @param {string} serviceName - The name of the service (e.g., 'accounts-subgraph')
 * @returns {NodeSDK} The initialized OpenTelemetry SDK
 */
function initializeOpenTelemetry(serviceName) {
  const environment = process.env.ENVIRONMENT || 'demo';
  const serviceVersion = process.env.SERVICE_VERSION || '1.0.0';
  const authToken = process.env.DASH0_AUTH_TOKEN;
  const tracesEndpoint = process.env.DASH0_TRACES_ENDPOINT;
  const metricsEndpoint = process.env.DASH0_METRICS_ENDPOINT;

  // Validate required configuration
  if (!authToken || !tracesEndpoint || !metricsEndpoint) {
    console.warn('⚠️  OpenTelemetry configuration incomplete. Telemetry data will not be exported.');
    console.warn('   Required: DASH0_AUTH_TOKEN, DASH0_TRACES_ENDPOINT, DASH0_METRICS_ENDPOINT');
  }

  // Configure resource with service information
  const resource = new Resource({
    [ATTR_SERVICE_NAME]: serviceName,
    [ATTR_SERVICE_VERSION]: serviceVersion,
    'deployment.environment': environment,
  });

  // Configure trace exporter
  const traceExporter = new OTLPTraceExporter({
    url: tracesEndpoint,
    headers: {
      'Authorization': authToken,
    },
  });

  // Configure metric exporter
  const metricExporter = new OTLPMetricExporter({
    url: metricsEndpoint,
    headers: {
      'Authorization': authToken,
    },
  });

  // Initialize the OpenTelemetry SDK
  const sdk = new NodeSDK({
    resource: resource,
    traceExporter: traceExporter,
    metricReader: new PeriodicExportingMetricReader({
      exporter: metricExporter,
      exportIntervalMillis: 60000, // Export metrics every 60 seconds
    }),
    // Configure trace context propagation to extract parent spans from incoming requests
    // This enables the router's trace context (traceparent header) to be properly linked
    textMapPropagator: new W3CTraceContextPropagator(),
    instrumentations: [
      getNodeAutoInstrumentations({
        // Enable only the instrumentations we need
        '@opentelemetry/instrumentation-http': {
          enabled: true,
        },
        '@opentelemetry/instrumentation-express': {
          enabled: true,
        },
        '@opentelemetry/instrumentation-graphql': {
          enabled: true,
        },
        '@opentelemetry/instrumentation-dns': {
          enabled: false,
        },
        '@opentelemetry/instrumentation-fs': {
          enabled: false,
        },
      }),
    ],
  });

  // Start the SDK
  sdk.start();
  console.log(`🔭 OpenTelemetry instrumentation initialized for ${serviceName}`);

  // Graceful shutdown
  process.on('SIGTERM', () => {
    sdk.shutdown()
      .then(() => console.log('OpenTelemetry SDK shut down successfully'))
      .catch((error) => console.log('Error shutting down OpenTelemetry SDK', error))
      .finally(() => process.exit(0));
  });

  return sdk;
}

module.exports = { initializeOpenTelemetry };

File: subgraphs/products-py/otel.py
"""
OpenTelemetry initialization for Python subgraph.
Provides manual span creation for GraphQL operations.
Includes trace context propagation for distributed tracing.
"""
import os
from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator
from opentelemetry.propagate import set_global_textmap


def initialize_opentelemetry(service_name: str):
    """
    Initialize OpenTelemetry with core SDK only.
    Uses manual span creation in GraphQL resolvers for precise control.

    Args:
        service_name: The name of the service (e.g., 'products-subgraph-py')
    """
    environment = os.getenv('ENVIRONMENT', 'demo')
    service_version = os.getenv('SERVICE_VERSION', '1.0.0')
    auth_token = os.getenv('DASH0_AUTH_TOKEN')
    traces_endpoint = os.getenv('DASH0_TRACES_ENDPOINT')

    # Validate required configuration
    if not auth_token or not traces_endpoint:
        print('⚠️  OpenTelemetry configuration incomplete. Traces will not be exported.')
        print('   Required: DASH0_AUTH_TOKEN, DASH0_TRACES_ENDPOINT')

    # Configure resource with service information
    resource = Resource.create({
        "service.name": service_name,
        "service.version": service_version,
        "deployment.environment": environment,
    })

    # Configure trace exporter
    headers = {"Dash0-Dataset": "gtm-dash0"}
    if auth_token:
        headers["Authorization"] = auth_token
    trace_exporter = OTLPSpanExporter(
        endpoint=traces_endpoint if traces_endpoint else "http://localhost:4317/v1/traces",
        headers=headers,
    )

    # Configure tracer provider with trace exporter
    tracer_provider = TracerProvider(resource=resource)
    tracer_provider.add_span_processor(BatchSpanProcessor(trace_exporter))
    trace.set_tracer_provider(tracer_provider)

    # Configure trace context propagation to extract parent spans from incoming requests
    # This enables the router's trace context (traceparent header) to be properly linked
    propagator = TraceContextTextMapPropagator()
    set_global_textmap(propagator)

    print(f'🔭 OpenTelemetry initialized for {service_name}')

    return tracer_provider
